### DISCUSSION TITLE
Pg_rewind network failure

### DISCUSSION THREAD

**Discussion by lenow55** [cloudnative-pg/cloudnative-pg#2119] [Q&A] [answered]
## Description
I have two nodes with two instances postgresql. When i shutdown node with primary instance, replica becomes new primary. After that i turn on node and old primary becomes replica. New replica tries to execute pg_rewind and this process fail with data corruption.
**pg_rewind on new replica**
`"ts":"2023-03-27T08:04:05Z","msg":"Starting up pg_rewind","logging_pod":"cluster-example-6","pgdata":"/var/lib/postgresql/data/pgdata","options":["-P","--source-server","host=cluster-example-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=cluster-example-6 sslmode=verify-ca dbname=postgres","--target-pgdata","/var/lib/postgresql/data/pgdata","--debug","--restore-target-wal"]`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"pg_rewind: using for rewind restore_command = '/controller/manager wal-restore --log-destination /controller/log/postgres.json %f %p'"`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"pg_rewind: connected to server"`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"pg_rewind: executing \"/usr/lib/postgresql/15/bin/postgres\" for target server to complete crash recovery"`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"2023-03-27 08:04:05.691 UTC [25] LOG:  database system was interrupted; last known up at 2023-03-27 08:01:05 UTC"`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"2023-03-27 08:04:05.692 UTC [25] LOG:  database system was not properly shut down; automatic recovery in progress"`
....
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"pg_rewind: servers diverged at WAL location 0/34000110 on timeline 16"`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"pg_rewind: rewinding from last common checkpoint at 0/34000060 on timeline 16"`
`"ts":"2023-03-27T08:04:05Z","logger":"pg_rewind","msg":"pg_rewind: reading source file list"`
`"ts":"2023-03-27T08:04:06Z","logger":"pg_rewind","msg":"pg_rewind: reading target file list"`
`"ts":"2023-03-27T08:04:06Z","logger":"pg_rewind","msg":"pg_rewind: reading WAL in target"`
....
`"ts":"2023-03-27T08:04:11Z","logger":"pg_rewind","msg":"pg_rewind: received chunk for file \"pg_wal/0000000B000000000000002B.partial\", offset 9437184, size 1048576"`
`"ts":"2023-03-27T08:04:11Z","logger":"pg_rewind","msg":"pg_rewind: received chunk for file \"pg_wal/0000000B000000000000002B.partial\", offset 10485760, size 1048576"`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"pg_rewind: error: unexpected result while fetching remote files: server closed the connection unexpectedly"`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"\tThis probably means the server terminated abnormally"`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"\tbefore or while processing the request."`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"server closed the connection unexpectedly"`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"\tThis probably means the server terminated abnormally"`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"\tbefore or while processing the request."`
`"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"backend> "`
_Full log on github_
https://github.com/lenow55/cloud_native/blob/master/mon_con/log10kill/cl6-prim2repl-fail-pod-kill-min-startup.json
**And logs on new primary**
`"ts":"2023-03-27T08:04:17Z","logger":"postgres","msg":"record","logging_pod":"cluster-example-1","record":{"log_time":"2023-03-27 08:04:17.109 UTC","user_name":"streaming_replica","database_name":"postgres","process_id":"2506","connection_from":"10.40.1.54:59946","session_id":"64214df5.9ca","session_line_num":"1","command_tag":"SELECT","session_start_time":"2023-03-27 08:04:05 UTC","virtual_transaction_id":"3/189","transaction_id":"0","error_severity":"LOG","sql_state_code":"08006","message":"could not send data to client: Connection reset by peer","query":"SELECT path, begin,\n  pg_read_binary_file(path, begin, len, true) AS chunk\nFROM unnest ($1::text[], $2::int8[], $3::int4[]) as x(path, begin, len)","application_name":"cluster-example-6","backend_type":"client backend","query_id":"0"}`
`"ts":"2023-03-27T08:04:17Z","logger":"postgres","msg":"record","logging_pod":"cluster-example-1","record":{"log_time":"2023-03-27 08:04:17.109 UTC","user_name":"streaming_replica","database_name":"postgres","process_id":"2506","connection_from":"10.40.1.54:59946","session_id":"64214df5.9ca","session_line_num":"2","command_tag":"SELECT","session_start_time":"2023-03-27 08:04:05 UTC","virtual_transaction_id":"3/189","transaction_id":"0","error_severity":"FATAL","sql_state_code":"08006","message":"connection to client lost","query":"SELECT path, begin,\n  pg_read_binary_file(path, begin, len, true) AS chunk\nFROM unnest ($1::text[], $2::int8[], $3::int4[]) as x(path, begin, len)","application_name":"cluster-example-6","backend_type":"client backend","query_id":"0"}`
_Full log on github_
https://github.com/lenow55/cloud_native/blob/master/mon_con/log10kill/cl1-repl2prim-fail-pod-kill-min.json

## Versions
- cloudnative_pg: v1.19.1
- kubernetes: v1.26.1
- postgresql: 15.2

## To reproduce
1. Setup fresh pg cluster by simple example on two nodes (two is enough)
2. Trigger failover by deleting primary pod
3. Ensure that you have on new primary less data than new replica
4. Restart node with new primary
5. Check logs on new replica

## Cause of fail
Replica connects to new leader by using kubernetes service -rw and pg_rewind uses this connection to download data from leader. During this process, the leader's address is replaced with the service address.
`cluster-example-rw    ClusterIP   10.103.179.133   <none>        5432/TCP   29h`
`cluster-example-1       10.40.1.54               replica`
`cluster-example-2       10.40.3.255              primary`
However, at an indefinite time, **kube-proxy does not convert the pod address to the service address**.
![Screenshot_20230329_231217](https://user-images.githubusercontent.com/99322377/228656545-20d24491-5b43-422f-918f-64a2ffc7395b.png)
The replica listens to the connection from the service, but a packet with a different address arrives on the same port and therefore the connection is reset. _Tcp packet with RST flag_ And pg_rewind fail.

## Solution
Using CurrentPrimary pod name with service -any to get direct access to primary instance. I realised it in my project fork, but i don't know how to make correct pullRequest.

Please help me to understand this situation with kubernetes servise. May be i use wrong cluster configuration. _Sorry for my bad english_

**Comment by sxd** [comment]
Hi @lenow55 

Can you confirm that the data on the node isn't actually corrupted? because we can see this error:

```
"ts":"2023-03-27T08:04:11Z","logger":"pg_rewind","msg":"pg_rewind: received chunk for file \"pg_wal/0000000B000000000000002B.partial\", offset 9437184, size 1048576"
"ts":"2023-03-27T08:04:11Z","logger":"pg_rewind","msg":"pg_rewind: received chunk for file \"pg_wal/0000000B000000000000002B.partial\", offset 10485760, size 1048576"
"ts":"2023-03-27T08:04:17Z","logger":"pg_rewind","msg":"pg_rewind: error: unexpected result while fetching remote files: server closed the connection unexpectedly"
```

Even if the connection is reset, the operator will try again, since it's a network error unless pg_rewind finds that there's actual data corrupted on the side that it's streaming the data.
The `-any` service was deprecated and was never created to be used for user, isn't a good idea to use that service.

If you don't have data corrupted, this looks like an issue with your Kubernetes network rather than an issue in the operator, either way, we may find something here, for now I'm turning this into a discussion and we can create an issue if we found exactly what was going on.

Best Regards!

**Reply by lenow55** [reply]
Hi @sxd, thank you for you attention.
On start up pg_rewind try to download data from new primary node.
As i understart firs check controll point makes correctly:
[https://github.com/lenow55/cloud_native/blob/1642250afb676bbcb773ab6c0c807b33a0dfd3ed/log12svc/cl2-prim2repl-fail-kill-startup.json#L28](url)
After that pg_rewind start streaming process, which is interrupted due to an error.
[https://github.com/lenow55/cloud_native/blob/1642250afb676bbcb773ab6c0c807b33a0dfd3ed/log12svc/cl2-prim2repl-fail-kill-startup.json#L770](url)
This error i described in previous post. You can find it from wireshark logs.
I'll try to explain it again.
Replica -> send request to service point 10.103.179.133 on 5432/TCP port.
Primary -> recive request from replica and send answer with his ip address 10.40.3.255 on 5432/TCP port
![image](https://github.com/cloudnative-pg/cloudnative-pg/assets/99322377/ec039d25-f829-4bc4-845b-bb0729391397)
**Kubernetes proxy pod substitute primary ip by service ip 10.40.3.255 -> 10.103.179.133**
![image](https://github.com/cloudnative-pg/cloudnative-pg/assets/99322377/4ad6082a-cc0e-49a9-89d8-d182e599a317)
Replica -> recive answer from service point.

But at some point **Kubernetes proxy pod do not substitute primary ip by service ip**
As a result:
Replica -> recive answer from **Primary IP** which is not match with Service IP
![Screenshot_20230526_194942](https://github.com/cloudnative-pg/cloudnative-pg/assets/99322377/b71bed0c-5c73-47e6-87e3-22d3632fcf39)
After that replica requests to stop sending packets from this address by packet with RST flag.
Primary -> recive RST send RST and close connection. (It seems like replica close connection)
Replica -> recive RST and close connection. (It seems like primary close connection)
**FAIL COMPLETE**
After fail pg_rewind try to do it again, but in streaming process pg_hba.conf file is corrupted. In result pg_ctl can't start instance correctly.
[https://github.com/lenow55/cloud_native/blob/1642250afb676bbcb773ab6c0c807b33a0dfd3ed/mon_con/log10kill/cl6-prim2repl-fail-pod-kill-min-startup.json#LL1019C180-L1020C187](url)

I guess that kube-proxy renew iptables rules with some period and this process has an impact on NAT.
![image](https://github.com/cloudnative-pg/cloudnative-pg/assets/99322377/b0b56746-1bca-45e4-abe1-10766f93c278)
I found these parameters in kube-proxy configuration. I know that i can increase these values, but if i have bigger data spliting than now, connection will lost. 

I suggest using direct ip access to primary pod excluding service, but we should get this ip before start pg_rewind process, and i use -any service to make it.

Best Regards!

**Reply by gbartolini** [reply]
I don't think `-any` could possibly work in case you have more than one standby. Am I seeing it right?

**Reply by lenow55** [reply]
Yes, i have two standby instances.
I'm not trying to say that the service `-any` is necessary.
I say there is a problem: the long streaming process in pg_rewind is interrupted due to a network error. The error occurs due to unstable operation of `nat` in `kube-proxy`.
Solution: Instance should get direct access the primary by its name. **Not `-rw` service.**
I modified cloud-nativepg code and used `-any` service to resolve primary pod name at ip address. I can't do same operation via `-rw` service, because it doesn't have `spec: publishNotReadyAddresses: true`.
You can see my changes here:
[https://github.com/lenow55/cloud_native_fork/commit/50a602b0784fd4061217ac549d9c873526fefc3b](url)
[https://github.com/lenow55/cloud_native_fork/commit/7cba555c359933544ad6fbe37284bd9e09118290](url)

**Comment by lenow55** [comment, marked as answer]
It seems that updating the operating system or packages solved this problem. I tried to reproduce this on the 1.27 version of kubernetes and couldn't do it. Thank you for your attention. If I find the problem again, I will try to describe it in more detail.
